component/xload/comp/lister.go

```diff
@@ -0,0 +1,198 @@
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in all
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE
+*/
+
+package comp
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/Azure/azure-storage-fuse/v2/common/config"
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/component/xload/common"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+)
+
+// verify that the below types implement the xcomponent interfaces
+var _ common.XComponent = &lister{}
+var _ common.XComponent = &remoteLister{}
+
+// verify that the below types implement the xenumerator interfaces
+var _ enumerator = &remoteLister{}
+
+const LISTER string = "lister"
+
+type lister struct {
+	common.XBase
+	path string // base path of the directory to be listed
+}
+
+type enumerator interface {
+	mkdir(name string) error
+}
+
+// --------------------------------------------------------------------------------------------------------
+
+type remoteLister struct {
+	lister
+	listBlocked bool
+}
+
+func NewRemoteLister(path string, remote internal.Component) (*remoteLister, error) {
+	log.Debug("lister::NewRemoteLister : create new remote lister for %s", path)
+
+	rl := &remoteLister{
+		lister: lister{
+			path: path,
+		},
+		listBlocked: false,
+	}
+
+	rl.SetName(LISTER)
+	rl.SetRemote(remote)
+	rl.Init()
+	return rl, nil
+}
+
+func (rl *remoteLister) Init() {
+	rl.SetThreadPool(common.NewThreadPool(common.MAX_LISTER, rl.Process))
+	if rl.GetThreadPool() == nil {
+		log.Err("remoteLister::Init : fail to init thread pool")
+	}
+}
+
+func (rl *remoteLister) Start() {
+	log.Debug("remoteLister::Start : start remote lister for %s", rl.path)
+	rl.GetThreadPool().Start()
+	rl.GetThreadPool().Schedule(&common.WorkItem{CompName: rl.GetName()})
+}
+
+func (rl *remoteLister) Stop() {
+	log.Debug("remoteLister::Stop : stop remote lister for %s", rl.path)
+	if rl.GetThreadPool() != nil {
+		rl.GetThreadPool().Stop()
+	}
+	rl.GetNext().Stop()
+}
+
+// wait for the configured block-list-on-mount-sec to make the list call
+func waitForListTimeout() error {
+	var blockListSeconds uint16 = 0
+	err := config.UnmarshalKey("azstorage.block-list-on-mount-sec", &blockListSeconds)
```

Comment:
Why do we want to wait here for this timeout. Ideally customer going for prefetch shall not have any wait timeout given,

Reply:
If the 10 second timeout is default, then on enabling this prefetch feature that shall get disabled as whole idea is to download all the contents as fast as possible so waiting for 10 seconds to unblock does not make sense here.

Reply:
The wait logic is for cases when user has set `block-list-on-mount-sec` parameter in the config. The default value of this parameter is 0. So, customers won't be impacted if they don't set this parameter.
Another reason for adding this is because we call the List() method in azstorage component, which waits till this timeout before making the REST call.

Reply:
Ideally customers using prefetch shall not have this wait time configured so it shall be safe to assume its 0 and not wait here. We can discuss this with the team,

Reply:
Yes, the default value is 0. So, if not configured, users won't have to wait for this timeout